perceval/backends/gerrit.py

```diff
@@ -0,0 +1,297 @@
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+#
+# Authors:
+#   Alvaro del Castillo San Felix <acs@bitergia.com>
+#
+
+'''Gerrit backend for Perseval'''
+
+
+from datetime import datetime
+import json
+import logging
+import os.path
+import re
+import subprocess
+from time import time
+
+from ..backend import Backend, BackendCommand
+from ..cache import Cache
+from ..errors import BackendError, CacheError
+from ..utils import DEFAULT_DATETIME, str_to_datetime
+
+
+class Gerrit(Backend):
+
+    name = "gerrit"
+
+    def __init__(self, user=None, url=None, nreviews=None,
+                 cache=None, **nouse):
+        super().__init__(cache=cache)
+        self.repository = url
+        self.nreviews = nreviews
+        self.last_review = None  # last review processed
+        self.number_results = None  # last number of results
+        self.client = GerritClient(self.repository, user, nreviews)
+
+    def fetch_from_cache(self):
+        """Fetch the bugs from the cache.
+
+        It returns the bugs stored in the cache object provided during
+        the initialization of the object. If this method is called but
+        no cache object was provided, the method will raise a `CacheError`
+        exception.
+
+        :returns: a generator of items
+
+        :raises CacheError: raised when an error occurs accesing the
+            cache
+        """
+        if not self.cache:
+            raise CacheError(cause="cache instance was not provided")
+
+        cache_items = self.cache.retrieve()
+
+        while True:
+            try:
+                raw_items = next(cache_items)
+                if raw_items is None:
+                    break
+                reviews = self._parse_reviews(raw_items)
+            except StopIteration:
+                break
+
+            for item in reviews:
+                yield item
+
+    def fetch(self, from_date=DEFAULT_DATETIME):
+        self._purge_cache_queue()
+
+        self.last_item = None
+        reviews = self._get_reviews()
+
+        while reviews:
+            review = reviews.pop(0)
+            updated = datetime.fromtimestamp(review['lastUpdated'])
+            if updated <= from_date:
+                logging.debug("No more updates for %s" % (self.repository))
+                break
+
+            yield review
+
+            if not reviews:
+                reviews = self._get_reviews()
+
+    def _get_reviews(self):
+        reviews = []
+
+        if self.number_results and self.number_results < self.nreviews:
+            # No more reviews
+            return reviews
+
+        task_init = time()
+        raw_data = self.client.get_items(self.last_item)
+        self._push_cache_queue(raw_data)
+        self._flush_cache_queue()
+        reviews = self._parse_reviews(raw_data)
+        self.last_item = self.client.get_next_item(self.last_item, reviews[-1])
```

Comment:
I think `last_item` can be stored outside of this method. In `fetch()`, for instance, and then,`get_reviews` will receive `last_item` as a parameter.